plugins/metalink/utils.c

```diff
@@ -720,128 +720,171 @@ TDNFParseUrlTag(
-            {
-                dwError = TDNFParseUrlTag(ml_ctx, node);
-                BAIL_ON_TDNF_ERROR(dwError);
-            }
+            elementInfo->dwError = ERROR_TDNF_METALINK_PARSER_MISSING_FILE_SIZE;
+            pr_err("XML Parser Error:File size is missing: %s", size);
+            BAIL_ON_TDNF_ERROR(elementInfo->dwError);
         }
-        if (node->children) {
-            TDNFXmlParseData(ml_ctx, node->children, filename);
+        if(sscanf(size, "%ld", &(elementInfo->ml_ctx->size)) != 1)
+        {
+            elementInfo->dwError = ERROR_TDNF_INVALID_PARAMETER;
+            pr_err("XML Parser Warning: size is invalid value: %s\n", size);
+            BAIL_ON_TDNF_ERROR(elementInfo->dwError);
         }
-        node = node->next;
+    }
+    else if(!strcmp(elementInfo->startElement, TAG_NAME_HASH))
+    {
+        elementInfo->dwError = TDNFParseHashTag(userData, val, len);
+        BAIL_ON_TDNF_ERROR(elementInfo->dwError);
+    }
+    else if(!strcmp(elementInfo->startElement, TAG_NAME_URL) && len > 4)
+    {
+        elementInfo->dwError = TDNFParseUrlTag(userData, val, len);
+        BAIL_ON_TDNF_ERROR(elementInfo->dwError);
     }
 
 cleanup:
-    if(xmlContValue)
-    {
-        xmlFree(xmlContValue);
-        xmlContValue = NULL;
+    if(size != NULL){
+       TDNF_SAFE_FREE_MEMORY(size);
     }
-    return dwError;
+    return;
 error:
     goto cleanup;
 }
 
+void
+TDNFXmlParseEndElement(
+    void *userData,
+    const char *name
+    )
+{
+    struct MetalinkElementInfo* elementInfo = (struct MetalinkElementInfo*)userData;
+    elementInfo->endElement = name;
+
+    if(elementInfo->dwError != 0)
+       BAIL_ON_TDNF_ERROR(elementInfo->dwError);
+
+cleanup:
+     return;
+error:
+     goto cleanup;
+}
+
 uint32_t
 TDNFMetalinkParseFile(
     TDNF_ML_CTX *ml_ctx,
-    int fd,
+    FILE *file,
     const char *filename
     )
 {
     uint32_t dwError = 0;
-    xmlDoc *doc = NULL;
-    xmlNode *root_element = NULL;
+    struct stat st;
+    size_t file_size = 0;
+    char *buffer = NULL;
+    struct MetalinkElementInfo elementInfo;
+    elementInfo.startElement = NULL;
+    elementInfo.endElement = NULL;
+    elementInfo.dwError = 0;
 
-    if(!ml_ctx || (fd <= 0) || IsNullOrEmptyString(filename))
+    XML_Parser parser = XML_ParserCreate(NULL);
+    if(!ml_ctx || (file == NULL) || IsNullOrEmptyString(filename) || (parser == NULL))
     {
         dwError = ERROR_TDNF_INVALID_PARAMETER;
         BAIL_ON_TDNF_ERROR(dwError);
     }
 
-    //Read The File and get the doc object.
-    doc = xmlReadFd(fd, NULL, NULL, 0);
+    elementInfo.ml_ctx = ml_ctx;
+    elementInfo.filename = filename;
 
-    if (doc == NULL)
-    {
-        pr_err("%s: Error while reading xml from fd:%d \n", __func__, fd);
-        dwError = ERROR_TDNF_METALINK_PARSER_INVALID_DOC_OBJECT;
-        BAIL_ON_TDNF_ERROR(dwError);
+    XML_SetElementHandler(parser, TDNFXmlParseStartElement, TDNFXmlParseEndElement);
```

Comment:
These function has return type of XMLCALL. Please check once that is void or something else so can check if function failed.

Reply:
XML_SetElementHandler has return type void

```
void XMLCALL
XML_SetElementHandler(XML_Parser p,
                      XML_StartElementHandler start,
                      XML_EndElementHandler end);
```